### PR TITLE
Add Provider overload for DependencyHandlerScope invoke

### LINKED ISSUE #24503: Kotlin Dsl: operator fun NamedDomainObjectProvider<Configuration>.invoke
### Expected Behavior
```kotlin
val foo: NamedDomainObjectProvider<Configuration> by configurations.existing

dependencies { // this is DependencyHandlerScope
    foo("app.foo:foo:1.2.3")
}
```

### Current Behavior
```kotlin
val foo: NamedDomainObjectProvider<Configuration> by configurations.existing

dependencies { // this is DependencyHandlerScope
    foo.name("app.foo:foo:1.2.3")
}
```

### Context
There is already an existing `String.invoke` extension available in `DependencyHandlerScope`.
If you want to lazily add a dependency to this configuration, you have to use the `configuration.name.invoke` (or use `get` first).
Instead, adding an overload would allow you to directly invoke the `NamedDomainObjectProvider<Configuration>` without getting the configuration first:
```kotlin
// in DependencyHandlerScope
operator fun NamedDomainObjectProvider<Configuration>.invoke(dependencyNotation: Any): Dependency? = name.invoke(dependencyNotation)
```

### PR DESCRIPTION
**Pull request by hfhbd**
Fixes #24503 

### Context
There is already an existing String.invoke extension available in DependencyHandlerScope.
If you want to lazily add a dependency to this configuration, you have to use the configuration.name.invoke (or use get first).
Instead, adding an overload would allow you to directly invoke the NamedDomainObjectProvider<Configuration> without getting the configuration first.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/GradleKotlinDslIntegrationTest.kt [hunk 1]
```diff
@@ -293,6 +294,25 @@ class GradleKotlinDslIntegrationTest : AbstractPluginIntegrationTest() {
...
+            """
+            val foo by configurations.registering
+
+            dependencies {
+              foo(files("fixture.jar"))
+            }
+            """
+        )
+
+        build("dependencies")
```
Comment (jbartok) [inline subprojects/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/GradleKotlinDslIntegrationTest.kt:313]: Shouldn't this test actually compute something based on `DeepThought` and check the result, to make sure the class is available?

Comment (hfhbd) [reply subprojects/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/GradleKotlinDslIntegrationTest.kt]: Okay, I improved the test logic.

**Comment by jbartok** [comment]
@bot-gradle test this

**Review by eskatos — COMMENTED** [review]
This is looking good.

For uniformity, the same pattern should be applied to `DependencyConstraintHandlerScope` before we merge this though.

It looks like the binary compatibility check is confused, the changes in `accepted-public-api-changes.json` file shouldn't have been necessary given the new functions are property annotated. Don't bother about this, we'll take a look.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyConstraintHandlerScope.kt [hunk 2]
```diff
@@ -70,7 +96,7 @@ private constructor(
      * @return The dependency constraint.
      * @see [DependencyConstraintHandler.add]
      */
-    operator fun Configuration.invoke(dependencyConstraintNotation: Any): DependencyConstraint? =
+    operator fun Configuration.invoke(dependencyConstraintNotation: Any): DependencyConstraint =
```
Comment (hfhbd) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyConstraintHandlerScope.kt:99]: I didn't remember changing the nullability myself or by some tasks (maybe linting?), but the `DependencyConstraintHandlerDelegate` always returns non nullable `DependencyConstraint` due to Kotlin's type specification. This is also true for other methods in this class if you switch from a `constraints` call to the delegate call, see line 74/100. This line does not use `constraints.add` but `add` using the delegate (which is in fact `constraints`, but written in Kotlin).
Open question: Do you accept such source breaking (not binary breaking) change and do you want this change for the other methods too?

Comment (eskatos) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyConstraintHandlerScope.kt:99]: We'll handle the nullability discrepancy separately.
Could you please revert this line and have the added function above return `DependencyConstraint?` too for uniformity?

Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyConstraintHandlerScope.kt:99]: :x: This is an unnecessary change, please revert

**Review by eskatos — CHANGES_REQUESTED** [review]
Thanks! This is almost ready to be merged.

I requested one minor change.

Moreover, the DCO check is failing. Could you please follow the [instructions](https://github.com/gradle/gradle/pull/25005/checks?check_run_id=14269412966) to fix this?

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyConstraintHandlerScope.kt (hunk 2 above)
Comment (hfhbd) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyConstraintHandlerScope.kt]: Reverted and fixed DCO.

**Comment by jbartok** [comment]
@bot-gradle test this

**Review by eskatos — APPROVED** [review]
:+1:

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Comment by eskatos** [comment]
Thank you @hfhbd, this will be in Gradle 8.3

**Comment by hfhbd** [comment]
Do you want a follow up issue regarding the return type?

**Comment by eskatos** [comment]
@hfhbd, follow up PR about the nullability
* https://github.com/gradle/gradle/pull/26308
